### DISCUSSION TITLE
Video jitter/shudder when clicking on timeline bar

### DISCUSSION THREAD

**Discussion by dataguy73301** [smplayer-dev/smplayer#753] [General]
Has anyone noticed a very brief jitter or shudder when clicking on the timeline bar to advance forward or move backwards in a video clip?  It's very brief, but does not happen when you use the mouse scroll wheel to advance back or forward.  

I'm using version 23.6.0.10157 on Windows 11 and it did not happen with the previous version.  Perhaps it's a hardware or smplayer setting/preference?

Thank you for any input.

**Comment by Raza-0007** [comment]
Yes I have noticed it too. The judder is caused when you drag the timeline with the mouse, as soon as you let go of the mouse, the timeline position shifts slightly causing the judder. It is a recent issue and was not present a few versions ago. I am not exactly sure when it started.

The settings under "Options>Preferences>Interface>Seeking" are all at their default, so it is not an incorrect settings issue.

I believe there is a bug in the "seek to position" algorithm, as the 10 sec short jump, and 1 min medium jump, is also not accurate. Sometime the short jump is 10 sec, at other times it is 14 sec, then 12 seconds etc. Same discrepancy with medium jump. This was never an issue before, and it happens on all the videos, so it is not dependent on a specific video format, video codec or video container.

@smplayer-dev, can you please look into this for us? Updating something has caused a regression somewhere.

Thanks,

~Raza

**Comment by smplayer-dev** [comment]
I included a development version of mpv in the Windows package because version 0.35.1 has a problem when playing videos from Youtube. So maybe it's a bug in that development version. 
You can get other mpv versions here:
https://sourceforge.net/projects/mpv-player-windows/files/

**Reply by Raza-0007** [reply]
I downgraded to 0.35.1 stable mpv as soon as I downloaded the new smplayer release. I do not use smplayer for watching YouTube, so the development built was pointless for me. 

So the judder is happening on mpv 0.35.1 with smplayer 23.6.0.

The judder is not present if I open the same video files directly on mpv 0.35.1.

The short jump discrepancy that I mentioned, where instead of 10 sec back/forward, the video jumps anywhere between 12-16 sec randomly is unfortunately is present in mpv 0.35.1 itself, so it is a mpv issue. I just never realized that it was doing this before, and only noticed while testing for this judder. **Edit:** I was able to correct this in smplayer by selecting "precise seeking" under "Options>Preferences>Interface>Seeking".

 The judder is not present on smplayer 22.7.0 (July 2022 release) with mpv 0.35.1. It was introduced somewhere after that.

In order to notice the judder I recommend raising the audio volume high, because when the timeline judders, so does the audio. You actually notice the audio judder more than the video judder.

**Edit**: Just wanted to add that the speed of the mouse click seems to have an effect on the judder. When you move the timeline with a normal mouse click speed you experience the judder, but if you do it with a quick mouse click, the judder is either minimized or not present at all.

~Raza

**Comment by dataguy73301** [comment]
Thank you for the input!

**Comment by smplayer-dev** [comment]
I think the judder problem is caused by this change: https://github.com/smplayer-dev/smplayer/pull/692
For the problem of seeking not being accurate, try enabling precise seeking.

**Reply by Raza-0007** [reply]
@smplayer-dev You are most likely right about this. I believe the following settings is conflicting with the above mentioned commit.

"Options>Preferences>Interface>Seeking>Behavior of time slider: Seek to position when released"

The judder happens when the slider is moved with a mouse to a different position on the timeline and then released. Smplayer's settings is telling the slider to seek to the new position when the mouse is released, but the above commit is telling it to wait to see if the user keeps the mouse pressed and wants to move the slider further. This conflict is resulting in a visible brief judder in both audio and video when the mouse is released.

The judder might be gone if smplayer setting is changed to "seek to position with the mouse press" instead of mouse release. But I do not know how to go about implementing it or testing it.

Thanks, for the precise seeking tip, it solved that issue for me.

**Comment by Raza-0007** [comment]
@smplayer-dev Any chance this timeline judder issue can be fixed? There is no rush so take your time. It is not a deal breaker, just annoying.

**Comment by smplayer-dev** [comment]
I can fix it by reverting that commit but I want to try first if it can be fixed without removing it.

**Comment by Elinron** [comment]
Registered on Github just to talk about this timeline-bar jitter issue.

@smplayer-dev  @Raza-0007 Do I lose anything significant by going back to older smplayer version? The changelog says new mpv version, does it make a big difference? 

Also, what is process of downgrading? just using the old exe file will do or i have to uninstall the new version and then install the older version.

**Reply by Raza-0007** [reply]
@Elinron What you will lose by downgrading are bug fixes and enhancements. If an older version was working fine for you and you did not notice or need any newer enhancements, then you can safely downgrade. The process of downgrading will require you to uninstall the current SMPlayer and install the older one. 

Keep in mind that SMPlayer is composed of two separate software, one is the SMPlayer front-end, and the other is the mpv player engine that actually plays all the media. You can always upgrade/downgrade one or the other without breaking anything.

For example, I personally use the portable version of SMPlayer, so I have flexibility in a way that if I want, I can use an older or a newer version of mpv player. All I have to do is to replace the contents of the smplayer/mpv folder with the contents of a portable version of mpv downloaded separately. You cannot do this with the installable version of SMPlayer without having to rebuild from source.

I have used SMPlayer 22.7.0 (July 2022) with mpv 0.35.1 without any issues, if you want you can do that and that will get rid of this judder, but then you will miss all the bug fixes in SMPlayer itself. A very important one for me was the next subtitle toggle key not working in v22.7.0, and it is significant enough that I will wait for @smplayer-dev to fix this judder bug instead.

**Reply by Elinron** [reply]
@Raza-0007  I downgraded to 22.7 version and it works fine, no more jitters. It was actually very easy to downgrade, there is no need to uninstall, the exe file gives the (Downgrade) / (Overwrite) options.

**Reply by Raza-0007** [reply]
@Elinron Good that you were able to resolve this issue. I have never installed SMPlayer on my system ever, so I was not aware that such an option existed. I prefer the portable version.

Do keep in mind that the mpv version bundled in with SMPlayer 22.7.0 is v0.34, released in Nov 2021. It lacks all the great improvements in mpv player since then, like the new gpu-next rendering engine etc. I have been using gpu-next with vulkan api for a while now and with some other custom mpv options, it gives me the best quality video possible on smplayer. 

But if you are satisfied with mpv 0.34, then that is all that mattes in the end.

Hopefully Ricardo, the smplayer-dev, will be able to fix this judder issue soon, and even upgrade mpv to version 0.36 that was released yesterday.

**Reply by Elinron** [reply]
@Raza-0007 Thanks for all the information on the latest mpv release, I am using this on a 16 inch LG Gram, so cannot discern much difference. I will upgrade to the next SMPlayer version whenever it is released.

**Comment by smplayer-dev** [comment]
The problem should be fixed in version 10159. You can find packages for Windows here (the artifacts at the bottom):
https://github.com/smplayer-dev/smplayer/actions/runs/5654944085

**Reply by Raza-0007** [reply]
@smplayer-dev Thanks, the commit revert has fixed the issue nicely. 

However, I do understand what @BenjaminRi was trying to achieve with his commit https://github.com/smplayer-dev/smplayer/pull/692, to bring smplayer in line with how the other media players behave. But unfortunately it did not work as intended. 

And thanks, for compiling the new version for us. I still cannot get my head around trying to compile from source at GitHub myself. I tried a couple of times to compile mpv player in the past, but failed. I assume I will fail at compiling smplayer as well, so I depend on a pre-compiled version. :grin:

**Reply by BenjaminRi** [reply]
My apologies. I've been using a patched SMPlayer with this commit for months and I hadn't noticed this bug, though now I actually see what you mean. I think that the feature of being able to drag the slider from anywhere on the bar is really important, so maybe I'll look into fixing this issue when I've got time.

**Reply by BenjaminRi** [reply]
Interestingly, the bug does not occur in the Skinnable GUI. It's only the other GUI that is affected. So the partial revert makes sense.

**Comment by BenjaminRi** [comment]
The reason for the jumping slider is that the slider has limited resolution (usually `#define SEEKBAR_RESOLUTION 1000`) and whenever the slider is moved, it makes a round-trip to a time calculation (`int jump_time = (int) mdat.duration * value / SEEKBAR_RESOLUTION;`) and is rounded to another value, which is written back to the timeline bar and makes the seeker jump. Even with the reverted commit, you can still see a little jump of the slider whenever you click on the timeline bar.

It seems to me like this issue has to be resolved first, and then dragging from anywhere can be made to work again.

**Reply by Raza-0007** [reply]
@BenjaminRi Thank you for taking the time to work on this issue. I also noticed the slider jump, and it is still present with the commit reverted, but the judder is gone. So the slider jump might be playing a factor in causing the judder or it could be something else. 

You mentioned earlier that the judder is not present in the sinkable GUI. Is the timeline slider jump present in that GUI? If it is then it has to be something else that is causing the judder. Maybe comparing the sinkable GUI code with the base GUI code will help. 

Unfortunately, since I cannot compile smplayer myself, I cannot change code and test, so this is all the help I can offer. Hopefully the issue will be identified and fixed, as I do like the ability to press the slider and move it across the timeline.

**Reply by BenjaminRi** [reply]
If you are on Ubuntu 22.04 it's very easy to compile:

```sh
git clone https://github.com/smplayer-dev/smplayer.git
cd smplayer
sudo apt install zlib1g-dev fakeroot build-essential devscripts debhelper g++
sudo apt install qtbase5-dev qt5-qmake qtscript5-dev qttools5-dev-tools qtbase5-private-dev libqt5webkit5-dev
make
# Run the player for testing:
./src/smplayer
```

It's called "Skinnable GUI". Able to skin, i.e. able to apply a custom theme. The Skinnable GUI uses a different code for the sliders.

> Is the timeline slider jump present in that GUI?

It's not.

> Maybe comparing the sinkable GUI code with the base GUI code will help.

Yep, that's what I'm analyzing. Unfortunately, it's not that simple because they are very different. Unfortunately, the code seems to be quite a bit more complex than necessary in this area.

**Reply by BenjaminRi** [reply]
> Hopefully the issue will be identified and fixed, as I do like the ability to press the slider and move it across the timeline.

You can use the Skinnable GUI if you like this feature. @smplayer-dev did not remove the feature there. `Skinnable GUI` can be selected in `Options` -> `Preferences` -> `Interface` -> `GUI` dropdown. But the menu point only exists if you have a theme installed, e.g. on Linux that is in `$HOME/.config/smplayer/themes/`.

**Reply by Raza-0007** [reply]
@BenjaminRi, I am actually a Windows guy. Though I have used linux occasionally, I don't have mastery over it. In Windows you need to install various software and their dependencies to compile, at least that was how it was for mpv player, I have not tried to compile smplayer.

As for the GUI, I have been using smplayer's base GUI for a very long time. It is heavily customized for my needs and set exactly how I like it. So for me switching to the any other GUI will diminished the experience.

If this timeline issue is complicated, then take your time, there is absolutely no need to rush. It will most likely increase your knowledge in video player coding, if you are into coding that is.

**Comment by BenjaminRi** [comment]
I think I was able to isolate the root cause of the jitter when clicking on the timeline bar with the 'seek anywhere' feature:

```cpp
void TimeSlider::mouseReleased() {
	#if DEBUG
	printf("TimeSlider::mouseReleased\n");
	#endif
	//emit posChanged( value() );
}
```

The seek operation is applied twice: Once on mouse press, once on mouse release. Since between mouse press and release, a couple dozen milliseconds elapsed, the video already continued playing and will be reset to the previous position again, leading to this jitter. You can reproduce this nicely by holding the left mouse button longer and see the video jump back on release.

I believe I can fix this and it will also fix the jumping slider as an added bonus.

**Reply by Raza-0007** [reply]
@BenjaminRi, Thats brilliant! I also noticed the effect of the mouse click speed on the judder. I mentioned it in one of the posts earlier that the judder was either minimized or even absent with a quick mouse click. Your discovery seems to confirm that.

Also check the setting under "Options>Preferences>Interface>Seeking>Behavior of time slider:Seek to position when released". I believe this setting might conflict with your commit too, as this setting is telling smplayer to seek to position with mouse release, but your commit tells smplayer to wait and see if the user keeps the mouse pressed and wants to move the slider further. 

Just my 2 cents.

**Reply by BenjaminRi** [reply]
> Also check the setting under "Options>Preferences>Interface>Seeking>Behavior of time slider:Seek to position when released". I believe this setting might conflict with your commit too

Good point. I would have ran right into the next bug. OK it is becoming more complicated :rofl: but I'm not yet out of ideas.

**Comment by smplayer-dev** [comment]
New build for Windows with the new PR #771 by BenjaminRi:
https://github.com/smplayer-dev/smplayer/actions/runs/5706681481

Could you verify that this version doesn't have this judder problem?

**Reply by Raza-0007** [reply]
The judder is gone! Yay!!! :smiley:

The slider still jumps backwards a little when you release the mouse, but it does not cause any judder in the video or the audio.

During testing I did notice that when you move the slider to the different position in the timeline, release the mouse, and without changing the mouse position, immediately click the slider again to move it to a different position, it does cause a brief skip in the video. I believe this skip is most likely caused by the fact that the slider had shifted slightly backwards when you released the mouse, and has to move back to the initial mouse release position when you immediately click it again. 

I wonder what causes the slider to shift slightly backwards, is it trying to seek to a precise time or frame?

In any case, it does not cause any judder, just a skip, and it is not something serious, but if @BenjaminRi has some free time, maybe he can have a look into why this happens.

All thanks to @BenjaminRi for taking the time to fix this issue, it certainly has made an improvement in SMPlayer GUI.

**Reply by BenjaminRi** [reply]
> The slider still jumps backwards a little when you release the mouse

Unfortunatey, this cannot be fixed easily. The problem is really that the slider has finite precision and seeks the video which then rounds to seconds and sends that back to the slider. This round-trip is very unfortunate in its implementation and the entire thing would require a deeper rework to work out all the kinks. Additionally, I believe there is an inaccuracy in how SMPlayer (or in general, Qt sliders) calculates the slider position. A lot of time could be spent further improving that code. There is also some legacy stuff in there like `ENABLE_DELAYED_DRAGGING` which I cannot for the life of me understand - why would anyone delay dragging, what's the point of this? I worry that if I were to clean it all up, I would break legitimate use-cases that I don't yet understand.

> and without changing the mouse position, immediately click the slider again to move it to a different position, it does cause a brief skip in the video.

Do you have "Seek to position when released" enabled? Looks like I can reproduce it with that setting, but not the other one. Is dragging and immediately clicking again a use-case, does that bother people? I'll think about what can be done about it.

**Reply by BenjaminRi** [reply]
Oh, I think I found a minor thing I missed that I could improve on. I believe it would fix this issue.

**Reply by Raza-0007** [reply]
>    Do you have "Seek to position when released" enabled? Looks like I can reproduce it with that setting, but not the other one. Is dragging and immediately clicking again a use-case, does that bother people? I'll think about what can be done about it.

Yes, "seek to position when released" is what is set by default. I never changed it because before your commit, one could not drag the timeline in smplayer, so this setting did not have much effect. But now this setting has become important. I can always switch to the "seek to position while dragging" option, as you are right, that does not cause this issue.

Immediately clicking on the timeline after moving to a new position is not a thing one would do normally, so you are right here too. If fixing this is too difficult and time consuming, then there is no need to spend time on it. I can live with it, and I am sure others can live with it as well.

**Reply by Raza-0007** [reply]
>Do you have "Seek to position when released" enabled? Looks like I can reproduce it with that setting, but not the other one.

Actually, I take it back. Even with the other setting the "seek to position while dragging", the issue is still there, it is just that you cannot notice it as much, because the video moves while dragging in any case. 

However, if you look closely at the slider when you release the mouse, it shifts backwards slightly, and if you click on the timeline immediately after releasing the mouse, the video skips back to the original release point. 

But again this is just nitpicking at the moment, and not worth spending too much of your time. But if you have discovered a quick fix, then it would be great if this is resolved.
